File: src/main/core/deps.js
'use strict';

const fs = require('node:fs');
const path = require('node:path');
const { JsonStore } = require('./store');

/**
 * Зависимости установленных модов (2.1).
 *
 * До 2.1 проверка сравнивала номер записи в реестре с зависимостями из
 * manifest.json. У модов Stardew с Nexus номер записи — «nexus:stardewvalley:1915»,
 * а в манифесте зависимость записана как «Pathoschild.ContentPatcher»: они не
 * совпадали никогда, и Content Patcher числился «недостающим», даже когда стоял.
 * Моды, поставленные руками, не учитывались вовсе, а требования Nexus
 * (Nautilus у модов Subnautica) не проверялись совсем.
 *
 * Теперь мод считается на месте, если совпало хоть что-то:
 *   • номер в реестре, в том числе номер Nexus без приставки;
 *   • UniqueID из manifest.json (Stardew) — и у модов ModHub, и у тех,
 *     что лежат в папке модов сами по себе;
 *   • имя папки или имя мода (для требований Nexus, у которых есть только имя).
 *
 * Для недостающего сразу ищется, откуда его поставить:
 *   • требование Nexus — это и есть номер на Nexus;
 *   • UniqueID Stardew — номер на Nexus спрашиваем у сайта SMAPI (smapi.io),
 *     ответ храним неделю;
 *   • зависимости Thunderstore и ModLinks — их же номера в каталоге.
 */

const SMAPI_API = 'https://smapi.io/api/v3.0/mods';
const WEEK = 7 * 24 * 60 * 60 * 1000;

function norm(value) {
  return String(value ?? '')
    .toLowerCase()
    .replace(/\(.*?\)/g, '')
    .replace(/[^a-z0-9а-яё]+/gi, '');
}

/** UniqueID из manifest.json в папках модов (Stardew): и свои, и чужие. */
function manifestIds(modsDir) {
  const out = new Set();
  if (!modsDir || !fs.existsSync(modsDir)) return out;
  const visit = (dir, depth) => {
    let entries;
    try {
      entries = fs.readdirSync(dir, { withFileTypes: true });
    } catch {
      return;
    }
    for (const entry of entries) {
      if (entry.isFile() && entry.name.toLowerCase() === 'manifest.json') {
        try {
          const text = fs.readFileSync(path.join(dir, entry.name), 'utf8').replace(/^\uFEFF/, '');
          const data = JSON.parse(text.replace(/,\s*([}\]])/g, '$1'));
          if (data?.UniqueID) out.add(String(data.UniqueID).toLowerCase());
        } catch {
          /* битый манифест — не наш вопрос */
        }
      } else if (entry.isDirectory() && depth < 3 && !entry.name.startsWith('.')) {
        visit(path.join(dir, entry.name), depth + 1);
      }
    }
  };
  visit(modsDir, 0);
  return out;
}

/** Всё, по чему можно узнать установленный мод. */
function presentKeys(records, { modsDir, loaderIds = [] } = {}) {
  const ids = new Set(loaderIds.map((id) => String(id).toLowerCase()));
  const names = new Set();
  for (const record of records) {
    if (record.missing) continue;
    ids.add(String(record.id).toLowerCase());
    const nexus = /^nexus:[^:]+:(\d+)$/.exec(record.id);
    if (nexus) ids.add(`nexus#${nexus[1]}`);
    if (record.uniqueId) ids.add(String(record.uniqueId).toLowerCase());
    for (const extra of record.uniqueIds ?? []) ids.add(String(extra).toLowerCase());
    names.add(norm(record.name));
    names.add(norm(record.folder));
  }
  for (const id of manifestIds(modsDir)) ids.add(id);
  // Папки в модах, которых нет в реестре (поставлены руками): их имена тоже считаются.
  try {
    if (modsDir && fs.existsSync(modsDir)) {
      for (const entry of fs.readdirSync(modsDir, { withFileTypes: true })) names.add(norm(entry.name.replace(/\.dll$/i, '')));
    }
  } catch {
    /* папки нет — и ладно */
  }
  names.delete('');
  return { ids, names };
}

/**
 * Чего не хватает включённым модам.
 * @returns {Array<{mod: string, modId: string, missing: string, missingId: string, kind: string, resolve: object|null}>}
 */
function findMissing(records, { modsDir, loaderIds = [], hide = [] } = {}) {
  const present = presentKeys(records, { modsDir, loaderIds });
  const hidden = new Set(hide.map(String));
  const out = [];
  const seen = new Set();
  for (const record of records) {
    if (record.missing || record.enabled === false) continue;
    // Зависимости из манифеста мода: UniqueID (Stardew), «Автор-Пакет» (Thunderstore), имя (ModLinks).
    for (const dep of record.dependencies ?? []) {
      const id = String(dep ?? '').trim();
      if (!id) continue;
      const key = id.toLowerCase();
      if (present.ids.has(key) || present.names.has(norm(id))) continue;
      if (/-bepinexpack(_\w+)?$/i.test(id) || /^smapi$/i.test(id) || key === 'pathoschild.smapi') continue;
      if (seen.has(`${record.id}|${key}`)) continue;
      seen.add(`${record.id}|${key}`);
      out.push({ mod: record.name, modId: record.id, missing: id, missingId: id, kind: 'manifest', resolve: null });
    }
    // Требования со страницы Nexus: номер и имя.
    for (const req of record.requires ?? []) {
      const id = String(req?.id ?? '');
      if (!id || hidden.has(id)) continue;
      if (present.ids.has(`nexus#${id}`) || present.names.has(norm(req.name))) continue;
      if (seen.has(`${record.id}|nexus#${id}`)) continue;
      seen.add(`${record.id}|nexus#${id}`);
      out.push({ mod: record.name, modId: record.id, missing: req.name || `#${id}`, missingId: id, kind: 'nexus', resolve: { id } });
    }
  }
  return out;
}

/**
 * UniqueID модов Stardew → номер на Nexus, по сайту SMAPI.
 * Ответ сайта держим неделю: и лишний раз не спрашиваем, и без сети помним.
 */
class SmapiIndex {
  constructor({ file, fetch: fetchImpl = globalThis.fetch, now = () => Date.now(), version = '2.1' } = {}) {
    this.store = new JsonStore(file, { items: {} });
    this.fetch = fetchImpl;
    this.now = now;
    this.version = version;
  }

  cached(id) {
    const item = this.store.data.items[String(id).toLowerCase()];
    return item && this.now() - item.at < WEEK ? item : null;
  }

  /** @returns {Promise<Record<string, {nexusId: string|null, name: string|null}>>} */
  async lookup(uniqueIds) {
    const wanted = [...new Set(uniqueIds.map((id) => String(id)))].filter(Boolean);
    const out = {};
    const ask = [];
    for (const id of wanted) {
      const hit = this.cached(id);
      if (hit) out[id] = hit;
      else ask.push(id);
    }
    if (ask.length) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), 15000);
      try {
        const response = await this.fetch(SMAPI_API, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json', 'User-Agent': `ModLaunch/${this.version}` },
          body: JSON.stringify({
            mods: ask.slice(0, 50).map((id) => ({ id, updateKeys: [] })),
            apiVersion: '4.0.0',
            gameVersion: '1.6.15',
            platform: 'Windows',
            includeExtendedMetadata: true,
          }),
          signal: controller.signal,
        });
        const list = response.ok ? await response.json() : [];
        for (const item of Array.isArray(list) ? list : []) {
          const id = String(item?.id ?? '');
          if (!id) continue;
          const meta = item.metadata ?? {};
          const nexusId = meta.nexusID ?? meta.nexusId ?? null;
          const entry = { nexusId: nexusId ? String(nexusId) : null, name: meta.name ?? null, at: this.now() };
          this.store.data.items[id.toLowerCase()] = entry;
          out[id] = entry;
        }
        this.store.save();
      } catch {
        /* сайт SMAPI недоступен — поставить недостающее можно будет руками */
      } finally {
        clearTimeout(timer);
      }
    }
    return out;
  }
}

module.exports = { findMissing, presentKeys, manifestIds, SmapiIndex, norm };

File: desktop/ModLaunch/Features/Deps.cs
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using ModLaunch.Core;
using ModLaunch.Games;
using ModLaunch.Mods;
using ModLaunch.Sources;

namespace ModLaunch.Features;

/// <summary>Недостающая зависимость: кому нужна, чего нет и как это поставить из каталога.</summary>
public sealed class Missing
{
    public required string Mod { get; init; }
    public required string ModId { get; init; }
    public required string Name { get; set; }
    public required string Id { get; init; }
    public required string Kind { get; init; } // manifest | nexus
    public string? ResolveId { get; set; }     // номер в каталоге этой игры
}

/// <summary>
/// Зависимости модов. Что «есть» — считаем щедро: номер в списке, UniqueID из
/// manifest.json любого мода в папке, имя папки. Так мод, поставленный руками,
/// тоже закрывает зависимость.
/// </summary>
public static partial class Deps
{
    [GeneratedRegex(@"\(.*?\)")] private static partial Regex Brackets();
    [GeneratedRegex(@"[^a-z0-9а-яё]+", RegexOptions.IgnoreCase)] private static partial Regex NotWord();
    [GeneratedRegex(@"^nexus:[^:]+:(\d+)$")] private static partial Regex NexusRecord();
    [GeneratedRegex(@"-bepinexpack(_\w+)?$", RegexOptions.IgnoreCase)] private static partial Regex LoaderPack();

    public static string Norm(string? value) => NotWord().Replace(Brackets().Replace((value ?? "").ToLowerInvariant(), ""), "");

    static HashSet<string> ManifestIds(string modsDir)
    {
        var ids = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
        if (!Directory.Exists(modsDir)) return ids;
        void Visit(string dir, int depth)
        {
            IEnumerable<string> entries;
            try { entries = Directory.EnumerateFileSystemEntries(dir).ToList(); } catch { return; }
            foreach (var entry in entries)
            {
                var name = Path.GetFileName(entry);
                if (File.Exists(entry) && name.Equals("manifest.json", StringComparison.OrdinalIgnoreCase))
                {
                    try
                    {
                        var text = Regex.Replace(File.ReadAllText(entry).TrimStart('﻿'), @",\s*([}\]])", "$1");
                        if (JsonNode.Parse(text).Str("UniqueID") is string id) ids.Add(id);
                    }
                    catch { }
                }
                else if (Directory.Exists(entry) && depth < 3 && !name.StartsWith('.')) Visit(entry, depth + 1);
            }
        }
        Visit(modsDir, 0);
        return ids;
    }

    public sealed record Present(HashSet<string> Ids, HashSet<string> Names);

    public static Present PresentKeys(IEnumerable<JsonObject> records, string modsDir, IEnumerable<string> loaderIds)
    {
        var ids = new HashSet<string>(loaderIds, StringComparer.OrdinalIgnoreCase);
        var names = new HashSet<string>();
        foreach (var r in records)
        {
            if (r.Bool("missing")) continue;
            var id = r.Str("id") ?? "";
            ids.Add(id);
            if (NexusRecord().Match(id) is { Success: true } m) ids.Add("nexus#" + m.Groups[1].Value);
            if (r.Str("uniqueId") is string u) ids.Add(u);
            names.Add(Norm(r.Str("name")));
            names.Add(Norm(r.Str("folder")));
        }
        ids.UnionWith(ManifestIds(modsDir));
        try
        {
            if (Directory.Exists(modsDir))
                foreach (var entry in Directory.EnumerateFileSystemEntries(modsDir))
                    names.Add(Norm(Regex.Replace(Path.GetFileName(entry), @"\.dll$", "", RegexOptions.IgnoreCase)));
        }
        catch { }
        names.Remove("");
        return new Present(ids, names);
    }

    public static List<Missing> Find(ModRegistry registry)
    {
        var game = registry.Game;
        var records = registry.List();
        var present = PresentKeys(records, registry.ModsDir, game.ThunderstorePackage is null ? [] : [game.ThunderstorePackage]);
        var hidden = game.NexusHide.ToHashSet();
        var seen = new HashSet<string>();
        var result = new List<Missing>();
        foreach (var r in records)
        {
            if (r.Bool("missing") || !r.Bool("enabled", true)) continue;
            var rid = r.Str("id") ?? "";
            var rname = r.Str("name") ?? rid;
            foreach (var dep in r.Arr("dependencies").Select(d => d?.ToString()?.Trim()).OfType<string>().Where(d => d != ""))
            {
                if (present.Ids.Contains(dep) || present.Names.Contains(Norm(dep))) continue;
                if (LoaderPack().IsMatch(dep) || dep.Equals("smapi", StringComparison.OrdinalIgnoreCase) || dep.Equals("Pathoschild.SMAPI", StringComparison.OrdinalIgnoreCase)) continue;
                if (!seen.Add($"{rid}|{dep.ToLowerInvariant()}")) continue;
                result.Add(new Missing { Mod = rname, ModId = rid, Name = dep, Id = dep, Kind = "manifest" });
            }
            foreach (var req in r.Arr("requires"))
            {
                var id = req.Str("id") ?? "";
                if (id == "" || hidden.Contains(id)) continue;
                if (present.Ids.Contains("nexus#" + id) || present.Names.Contains(Norm(req.Str("name")))) continue;
                if (!seen.Add($"{rid}|nexus#{id}")) continue;
                result.Add(new Missing { Mod = rname, ModId = rid, Name = req.Str("name") ?? "#" + id, Id = id, Kind = "nexus", ResolveId = id });
            }
        }
        return result;
    }

    /// <summary>Недостающее с номерами в каталоге: у Stardew UniqueID → номер Nexus через smapi.io.</summary>
    public static async Task<List<Missing>> FindResolved(ModRegistry registry, CancellationToken ct = default)
    {
        var missing = Find(registry);
        var manifest = missing.Where(m => m.Kind == "manifest").ToList();
        if (manifest.Count == 0) return missing;
        if (registry.Game.Catalog == CatalogKind.Nexus)
        {
            var found = await SmapiIndex.Lookup(manifest.Select(m => m.Id), ct);
            foreach (var m in manifest)
            {
                if (!found.TryGetValue(m.Id, out var hit) || hit.NexusId is null || registry.Game.NexusHide.Contains(hit.NexusId)) continue;
                m.ResolveId = hit.NexusId;
                if (hit.Name is not null) m.Name = hit.Name;
            }
        }
        else foreach (var m in manifest) m.ResolveId = m.Id;
        return missing;
    }

    /// <summary>Требования мода Nexus (и их требования), которых ещё нет, — самые глубокие первыми.</summary>
    public static async Task<List<(string Id, string Name)>> NexusPlan(GameDef game, ModRegistry registry, string modId, CancellationToken ct)
    {
        var present = PresentKeys(registry.List(), registry.ModsDir, []);
        var hidden = game.NexusHide.ToHashSet();
        var order = new List<(string, string)>();
        var visited = new HashSet<string> { modId };
        async Task Visit(string id, int depth)
        {
            if (depth > 3 || order.Count >= 12) return;
            List<(string Id, string Name)> reqs;
            try { reqs = (await Nexus.Details(game.NexusDomain!, game.NexusGameId, id, ct)).Requirements; } catch { return; }
            foreach (var (rid, name) in reqs)
            {
                if (hidden.Contains(rid) || !visited.Add(rid)) continue;
                if (present.Ids.Contains("nexus#" + rid) || present.Names.Contains(Norm(name))) continue;
                await Visit(rid, depth + 1);
                order.Add((rid, name));
            }
        }
        await Visit(modId, 0);
        return order;
    }
}

/// <summary>UniqueID мода Stardew → номер на Nexus (API smapi.io), с недельным кэшем.</summary>
public static class SmapiIndex
{
    static readonly JsonFile Cache = new(Path.Combine(Paths.DataDir, "smapi-index.json"), () => new JsonObject { ["items"] = new JsonObject() });

    public static async Task<Dictionary<string, (string? NexusId, string? Name)>> Lookup(IEnumerable<string> uniqueIds, CancellationToken ct)
    {
        var items = Cache.Data.Obj("items");
        var result = new Dictionary<string, (string?, string?)>(StringComparer.OrdinalIgnoreCase);
        var ask = new List<string>();
        foreach (var id in uniqueIds.Distinct(StringComparer.OrdinalIgnoreCase))
        {
            if (items[id.ToLowerInvariant()] is JsonObject hit && DateTime.UtcNow - DateTime.Parse(hit.Str("at") ?? "2000-01-01").ToUniversalTime() < TimeSpan.FromDays(7))
                result[id] = (hit.Str("nexusId"), hit.Str("name"));
            else ask.Add(id);
        }
        if (ask.Count == 0) return result;
        try
        {
            var body = new JsonObject
            {
                ["mods"] = new JsonArray(ask.Take(50).Select(id => (JsonNode)new JsonObject { ["id"] = id, ["updateKeys"] = new JsonArray() }).ToArray()),
                ["apiVersion"] = "4.0.0",
                ["gameVersion"] = "1.6.15",
                ["platform"] = "Windows",
                ["includeExtendedMetadata"] = true,
            };
            if (await Http.PostJson("https://smapi.io/api/v3.0/mods", body, ct: ct, timeoutSec: 15) is JsonArray list)
            {
                foreach (var item in list)
                {
                    var id = item.Str("id");
                    if (id is null) continue;
                    var meta = item?["metadata"];
                    var nexusId = meta?["nexusID"]?.ToString() ?? meta?["nexusId"]?.ToString();
                    var name = meta.Str("name");
                    items[id.ToLowerInvariant()] = new JsonObject { ["nexusId"] = nexusId, ["name"] = name, ["at"] = DateTime.UtcNow.ToString("o") };
                    result[id] = (nexusId, name);
                }
                Cache.Save();
            }
        }
        catch { /* сайт SMAPI недоступен — недостающее можно поставить руками */ }
        return result;
    }
}
